common_content_filter/src/DDSFilterField.cpp

```diff
@@ -0,0 +1,226 @@
+
+/**
+ * @file DDSFilterField.cpp
+ */
+
+#include "DDSFilterField.hpp"
+
+#include <rosidl_typesupport_introspection_c/identifier.h>
+#include <rosidl_typesupport_introspection_cpp/identifier.hpp>
+#include <rosidl_typesupport_introspection_c/message_introspection.h>
+#include <rosidl_typesupport_introspection_cpp/message_introspection.hpp>
+#include <rosidl_typesupport_introspection_cpp/field_types.hpp>
+#include <rosidl_runtime_c/string.h>
+
+#include <cassert>
+#include <unordered_set>
+#include <vector>
+
+#include "DDSFilterPredicate.hpp"
+#include "DDSFilterValue.hpp"
+#include "Log.hpp"
+
+namespace eprosima_common
+{
+namespace fastdds
+{
+namespace dds
+{
+namespace DDSSQLFilter
+{
+
+
+template<typename MembersType>
+bool
+DDSFilterField::get_msg_data_address(
+  const void * untype_members,
+  FieldAccessor & accessor,
+  const void * & data)
+{
+  const MembersType * members = static_cast<const MembersType *>(untype_members);
+  if (!members) {
+    return false;
+  }
+
+  const auto member = members->members_ + accessor.member_index;
+  // logDebug(DDSSQLFILTER, "DDSFilterField::get_msg_data_address "
+  //       << " accessor.member_index:" << accessor.member_index
+  //       << " accessor.array_index:" << accessor.array_index
+  //       << " member member->name:" << member->name_
+  //       << " member member->type_id_:" << static_cast<int>(member->type_id_)
+  //       << " member member->offset_:" << member->offset_
+  //       << " member member->is_array_:" << member->is_array_
+  //       << " member member->array_size_:" << member->array_size_
+  //       << " member member->is_upper_bound_:" << member->is_upper_bound_
+  //   );
+
+  uint64_t addr = reinterpret_cast<uint64_t>(data);
+  if (member->is_array_) {
+    size_t array_size = member->array_size_;
+    if (array_size == 0) {
+      array_size = member->size_function(
+        reinterpret_cast<void *>(addr + member->offset_));
+    }
+
+    if (accessor.array_index >= array_size) {
+      return false;
+    }
+
+    data = member->get_function(
+      reinterpret_cast<void *>(addr + member->offset_), accessor.array_index);
+  } else {
+    data = reinterpret_cast<void *>(addr + member->offset_);
+  }
+
+  return true;
+}
+
+bool DDSFilterField::set_value(
+  const void * data,
+  size_t n)
+{
+  bool last_step = access_path_.size() - 1 == n;
+  bool ret = false;
+  bool is_c_type_support;
+
+  // logDebug(
+  //   DDSSQLFILTER,
+  //   "DDSFilterField::set_value " << "access_path size:" << access_path_.size()
+  // );
+
+  const rosidl_message_type_support_t * type_support_introspection =
+    access_path_[n].type_support_introspection;
+  if (type_support_introspection->typesupport_identifier ==
+    rosidl_typesupport_introspection_c__identifier)
+  {
+    is_c_type_support = true;
+    ret = get_msg_data_address<rosidl_typesupport_introspection_c__MessageMembers>(
+      type_support_introspection->data, access_path_[n], data);
```

Comment:
Only need to deal with last one access_path?  

e.g. below case

a.b[3].c.d

a,b,c are all ros type message. Do we can access correct address ?
  
Based on my understanding, a,b,c,d should be corresponding to one access_path_ separately. `member_index` is the index of members and `array_index` is the index of array. 
So I think we should deal with each access_path_.

Reply:
> Only need to deal with last one access_path?

No.

> So I think we should deal with each access_path_.

The code does deal with each access_path_. Please notice that set_value calls the index begin with 0 as follows,
https://github.com/iuhilnehc-ynos/common_content_filter/blob/f727ad72306147156b4f4d640ac84d6c4fcecff1/common_content_filter/src/DDSFilterField.hpp#L113

and then it will recursive calling the set_value,
https://github.com/iuhilnehc-ynos/common_content_filter/blob/dffb8df81f961ce33a49c07c9cc375df2dd3b5b3/common_content_filter/src/DDSFilterField.cpp#L131

Reply:
Yes. Current code deals with all access_path_.  
For `access_path_.size() - 1 == n`, I misread it as `access_path_.size() - 1 = n`.